-1

**Failed Tests**:   Build 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9fa4a9/38717/summary.html
**COMMIT**: 33c945a8ba3c85e635a33cdf14130bc4857f82a9
**CMSSW**: CMSSW_14_1_X_2024-04-09-1100/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/44676/38717/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>Entering library rule at src/HeterogeneousTest/AlpakaDevice/plugins
>> Compiling alpaka/cuda src/HeterogeneousTest/AlpakaDevice/plugins/alpaka/AlpakaTestDeviceAdditionAlgo.dev.cc
>> Compiling alpaka/cuda edm plugin src/HeterogeneousTest/AlpakaDevice/plugins/alpaka/AlpakaTestDeviceAdditionModule.cc
>> Cuda Device Link tmp/el8_amd64_gcc12/src/HeterogeneousTest/AlpakaDevice/plugins/HeterogeneousTestAlpakaDevicePluginsCudaAsync/HeterogeneousTestAlpakaDevicePluginsCudaAsync_cudadlink.o 
nvlink error   : Undefined reference to '_ZN17alpaka_cuda_async4test13add_vectors_fERKN6alpaka22AccGpuUniformCudaHipRtINS1_9ApiCudaRtESt17integral_constantImLm1EEjEEPKfSA_Pfj' in 'tmp/el8_amd64_gcc12/src/HeterogeneousTest/AlpakaDevice/plugins/HeterogeneousTestAlpakaDevicePluginsCudaAsync/alpaka/AlpakaTestDeviceAdditionAlgo.dev.cc.o' (target: sm_60)
gmake: *** [tmp/el8_amd64_gcc12/src/HeterogeneousTest/AlpakaDevice/plugins/HeterogeneousTestAlpakaDevicePluginsCudaAsync/HeterogeneousTestAlpakaDevicePluginsCudaAsync_cudadlink.o] Error 255
>> Building alpaka/cuda edm plugin tmp/el8_amd64_gcc12/src/HeterogeneousTest/AlpakaDevice/plugins/HeterogeneousTestAlpakaDevicePluginsCudaAsync/libHeterogeneousTestAlpakaDevicePluginsCudaAsync.so
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02832/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/../lib/gcc/x86_64-redhat-linux-gnu/12.3.1/../../../../x86_64-redhat-linux-gnu/bin/ld.bfd: cannot find tmp/el8_amd64_gcc12/src/HeterogeneousTest/AlpakaDevice/plugins/HeterogeneousTestAlpakaDevicePluginsCudaAsync/HeterogeneousTestAlpakaDevicePluginsCudaAsync_cudadlink.o: No such file or directory
collect2: error: ld returned 1 exit status
gmake: *** [tmp/el8_amd64_gcc12/src/HeterogeneousTest/AlpakaDevice/plugins/HeterogeneousTestAlpakaDevicePluginsCudaAsync/libHeterogeneousTestAlpakaDevicePluginsCudaAsync.so] Error 1
Leaving library rule at src/HeterogeneousTest/AlpakaDevice/plugins
</pre>

--- a/HeterogeneousTest/AlpakaDevice/README.md
+++ b/HeterogeneousTest/AlpakaDevice/README.md
@@ -12,15 +12,13 @@ Alpaka-based libraries, and using them from multiple plugins.
 The package `HeterogeneousTest/AlpakaDevice` implements a library that defines and exports Alpaka
 device-side functions:
 ```c++
-namespace cms::alpakatest {
+namespace ALPAKA_ACCELERATOR_NAMESPACE::test {
 
-  template <typename TAcc>
-  ALPAKA_FN_ACC void add_vectors_f(TAcc const& acc, ...);
+  inline ALPAKA_FN_ACC void add_vectors_f(Acc1D const& acc, ...) { ... }
 
-  template <typename TAcc>
-  ALPAKA_FN_ACC void add_vectors_d(TAcc const& acc, ...);
+  inline ALPAKA_FN_ACC void add_vectors_d(Acc1D const& acc, ...) { ... }
 
-}  // namespace cms::alpakatest
+}  // namespace ALPAKA_ACCELERATOR_NAMESPACE::test
 ```
 
 The `plugins` directory implements the `AlpakaTestDeviceAdditionModule` `EDAnalyzer` that launches

--- a/HeterogeneousTest/AlpakaDevice/interface/alpaka/DeviceAddition.h
+++ b/HeterogeneousTest/AlpakaDevice/interface/alpaka/DeviceAddition.h
@@ -5,32 +5,31 @@
 
 #include <alpaka/alpaka.hpp>
 
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
 #include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
 
-namespace cms::alpakatest {
+namespace ALPAKA_ACCELERATOR_NAMESPACE::test {
 
-  template <typename TAcc>
-  ALPAKA_FN_ACC void add_vectors_f(TAcc const& acc,
-                                   float const* __restrict__ in1,
-                                   float const* __restrict__ in2,
-                                   float* __restrict__ out,
-                                   uint32_t size) {
+  inline ALPAKA_FN_ACC void add_vectors_f(Acc1D const& acc,
+                                          float const* __restrict__ in1,
+                                          float const* __restrict__ in2,
+                                          float* __restrict__ out,
+                                          uint32_t size) {
     for (auto i : cms::alpakatools::uniform_elements(acc, size)) {
       out[i] = in1[i] + in2[i];
     }
   }
 
-  template <typename TAcc>
-  ALPAKA_FN_ACC void add_vectors_d(TAcc const& acc,
-                                   double const* __restrict__ in1,
-                                   double const* __restrict__ in2,
-                                   double* __restrict__ out,
-                                   uint32_t size) {
+  inline ALPAKA_FN_ACC void add_vectors_d(Acc1D const& acc,
+                                          double const* __restrict__ in1,
+                                          double const* __restrict__ in2,
+                                          double* __restrict__ out,
+                                          uint32_t size) {
     for (auto i : cms::alpakatools::uniform_elements(acc, size)) {
       out[i] = in1[i] + in2[i];
     }
   }
 
-}  // namespace cms::alpakatest
+}  // namespace ALPAKA_ACCELERATOR_NAMESPACE::test
 
 #endif  // HeterogeneousTest_AlpakaDevice_interface_alpaka_DeviceAddition_h

--- a/HeterogeneousTest/AlpakaDevice/plugins/alpaka/AlpakaTestDeviceAdditionAlgo.dev.cc
+++ b/HeterogeneousTest/AlpakaDevice/plugins/alpaka/AlpakaTestDeviceAdditionAlgo.dev.cc
@@ -11,13 +11,12 @@
 namespace ALPAKA_ACCELERATOR_NAMESPACE::HeterogeneousTestAlpakaDevicePlugins {
 
   struct KernelAddVectorsF {
-    template <typename TAcc>
-    ALPAKA_FN_ACC void operator()(TAcc const& acc,
+    ALPAKA_FN_ACC void operator()(Acc1D const& acc,
                                   const float* __restrict__ in1,
                                   const float* __restrict__ in2,
                                   float* __restrict__ out,
                                   uint32_t size) const {
-      cms::alpakatest::add_vectors_f(acc, in1, in2, out, size);
+      test::add_vectors_f(acc, in1, in2, out, size);
     }
   };
 

--- a/HeterogeneousTest/AlpakaDevice/test/alpaka/testDeviceAddition.dev.cc
+++ b/HeterogeneousTest/AlpakaDevice/test/alpaka/testDeviceAddition.dev.cc
@@ -16,13 +16,12 @@
 using namespace ALPAKA_ACCELERATOR_NAMESPACE;
 
 struct KernelAddVectorsF {
-  template <typename TAcc>
-  ALPAKA_FN_ACC void operator()(TAcc const& acc,
+  ALPAKA_FN_ACC void operator()(Acc1D const& acc,
                                 const float* __restrict__ in1,
                                 const float* __restrict__ in2,
                                 float* __restrict__ out,
                                 uint32_t size) const {
-    cms::alpakatest::add_vectors_f(acc, in1, in2, out, size);
+    test::add_vectors_f(acc, in1, in2, out, size);
   }
 };
 

--- a/HeterogeneousTest/AlpakaKernel/README.md
+++ b/HeterogeneousTest/AlpakaKernel/README.md
@@ -12,19 +12,17 @@ Alpaka-based libraries, and using them from multiple plugins.
 The package `HeterogeneousTest/AlpakaKernel` implements a library that defines and exports Alpaka
 kernels that call the device functions defined in the `HeterogeneousTest/AlpakaDevice` library:
 ```c++
-namespace cms::alpakatest {
+namespace ALPAKA_ACCELERATOR_NAMESPACE::test {
 
   struct KernelAddVectorsF {
-    template <typename TAcc>
-    ALPAKA_FN_ACC void operator()(TAcc const& acc, ...) const;
+    ALPAKA_FN_ACC void operator()(Acc1D const& acc, ...) const { ... }
   };
 
   struct KernelAddVectorsD {
-    template <typename TAcc>
-    ALPAKA_FN_ACC void operator()(TAcc const& acc, ...) const;
+    ALPAKA_FN_ACC void operator()(Acc1D const& acc, ...) const { ... }
   };
 
-}  // namespace cms::alpakatest
+}  // namespace ALPAKA_ACCELERATOR_NAMESPACE::test
 ```
 
 The `plugins` directory implements the `AlpakaTestKernelAdditionModule` `EDAnalyzer` that launches

--- a/HeterogeneousTest/AlpakaKernel/interface/alpaka/DeviceAdditionKernel.h
+++ b/HeterogeneousTest/AlpakaKernel/interface/alpaka/DeviceAdditionKernel.h
@@ -5,13 +5,13 @@
 
 #include <alpaka/alpaka.hpp>
 
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
 #include "HeterogeneousTest/AlpakaDevice/interface/alpaka/DeviceAddition.h"
 
-namespace cms::alpakatest {
+namespace ALPAKA_ACCELERATOR_NAMESPACE::test {
 
   struct KernelAddVectorsF {
-    template <typename TAcc>
-    ALPAKA_FN_ACC void operator()(TAcc const& acc,
+    ALPAKA_FN_ACC void operator()(Acc1D const& acc,
                                   const float* __restrict__ in1,
                                   const float* __restrict__ in2,
                                   float* __restrict__ out,
@@ -21,8 +21,7 @@ namespace cms::alpakatest {
   };
 
   struct KernelAddVectorsD {
-    template <typename TAcc>
-    ALPAKA_FN_ACC void operator()(TAcc const& acc,
+    ALPAKA_FN_ACC void operator()(Acc1D const& acc,
                                   const double* __restrict__ in1,
                                   const double* __restrict__ in2,
                                   double* __restrict__ out,
@@ -31,6 +30,6 @@ namespace cms::alpakatest {
     }
   };
 
-}  // namespace cms::alpakatest
+}  // namespace ALPAKA_ACCELERATOR_NAMESPACE::test
 
 #endif  // HeterogeneousTest_AlpakaKernel_interface_alpaka_DeviceAdditionKernel_h

--- a/HeterogeneousTest/AlpakaKernel/plugins/alpaka/AlpakaTestKernelAdditionAlgo.dev.cc
+++ b/HeterogeneousTest/AlpakaKernel/plugins/alpaka/AlpakaTestKernelAdditionAlgo.dev.cc
@@ -16,7 +16,7 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::HeterogeneousTestAlpakaKernelPlugins {
                              float* __restrict__ out,
                              uint32_t size) {
     alpaka::exec<Acc1D>(
-        queue, cms::alpakatools::make_workdiv<Acc1D>(32, 32), cms::alpakatest::KernelAddVectorsF{}, in1, in2, out, size);
+        queue, cms::alpakatools::make_workdiv<Acc1D>(32, 32), test::KernelAddVectorsF{}, in1, in2, out, size);
   }
 
 }  // namespace ALPAKA_ACCELERATOR_NAMESPACE::HeterogeneousTestAlpakaKernelPlugins

--- a/HeterogeneousTest/AlpakaKernel/test/alpaka/testDeviceAdditionKernel.dev.cc
+++ b/HeterogeneousTest/AlpakaKernel/test/alpaka/testDeviceAdditionKernel.dev.cc
@@ -68,7 +68,7 @@ TEST_CASE("HeterogeneousTest/AlpakaKernel test", "[alpakaTestDeviceAdditionKerne
         // launch the 1-dimensional kernel for vector addition
         alpaka::exec<Acc1D>(queue,
                             cms::alpakatools::make_workdiv<Acc1D>(32, 32),
-                            cms::alpakatest::KernelAddVectorsF{},
+                            test::KernelAddVectorsF{},
                             in1_d.data(),
                             in2_d.data(),
                             out_d.data(),

--- a/HeterogeneousTest/AlpakaWrapper/src/alpaka/DeviceAdditionWrapper.dev.cc
+++ b/HeterogeneousTest/AlpakaWrapper/src/alpaka/DeviceAdditionWrapper.dev.cc
@@ -14,8 +14,7 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::test {
                              const float* __restrict__ in2,
                              float* __restrict__ out,
                              uint32_t size) {
-    alpaka::exec<Acc1D>(
-        queue, cms::alpakatools::make_workdiv<Acc1D>(32, 32), cms::alpakatest::KernelAddVectorsF{}, in1, in2, out, size);
+    alpaka::exec<Acc1D>(queue, cms::alpakatools::make_workdiv<Acc1D>(32, 32), KernelAddVectorsF{}, in1, in2, out, size);
   }
 
   void wrapper_add_vectors_d(Queue& queue,
@@ -23,8 +22,7 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::test {
                              const double* __restrict__ in2,
                              double* __restrict__ out,
                              uint32_t size) {
-    alpaka::exec<Acc1D>(
-        queue, cms::alpakatools::make_workdiv<Acc1D>(32, 32), cms::alpakatest::KernelAddVectorsD{}, in1, in2, out, size);
+    alpaka::exec<Acc1D>(queue, cms::alpakatools::make_workdiv<Acc1D>(32, 32), KernelAddVectorsD{}, in1, in2, out, size);
   }
 
 }  // namespace ALPAKA_ACCELERATOR_NAMESPACE::test

